app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```diff
@@ -0,0 +1,239 @@
+        String beforeBodySection = fileText.substring(0, startBodyIndex);
+        String bodySection = fileText.substring(startBodyIndex);
+        validateStringOrThrowException(bodySection);
+
+        // Swap out place-holder keywords for case property values within <body></body> section
+        bodySection = replace(bodySection, values);
+
+        // Write the new HTML to the desired  temp file location
+        // throws IO Exception
+        TemplatePrinterUtils.writeStringToFile(beforeBodySection + bodySection, outputPath);
+    }
+
+    /**
+     * Populate an input string with attribute keys formatted as {{ attr_key }}
+     * with attribute values.
+     * @param input String input
+     * @param values Bundle of String attribute key-value mappings
+     * @return The populated String
+     */
+    private static String replace(String input, Bundle values) {
+        // Split input into tokens bounded by {{ and }}
+        String[] tokens = TemplatePrinterUtils.splitKeepDelimiter(input, "\\{{2}", "\\}{2}");
+        for (int i=0; i<tokens.length; i++) {
+            String token = tokens[i];
+
+            // Every 2nd token is a attribute enclosed in {{ }}
+            if (i % 2 == 1) {
+
+                // Split token into tokenSplits bounded by < and >
+                String[] tokenSplits = TemplatePrinterUtils.splitKeepDelimiter(token, "<|(\\}{2})", ">|(\\{{2})");
+
+                // First and last tokenSplits are {{ and }}
+                for (int j=1; j<tokenSplits.length-1; j++) {
+                    String tokenSplit = tokenSplits[j];
+
+                    // tokenSplit is key or whitespace
+                    if (!tokenSplit.startsWith("<")) {
+
+                        // Remove whitespace from key
+                        String key = TemplatePrinterUtils.remove(tokenSplit, " ");
+
+                        if (values.containsKey(key) && (key = values.getString(key)) != null) {
+                            // Populate with value
+                            tokenSplits[j] = key;
+                        } else {
+                            // Empty if not found
+                            tokenSplits[j] = "";
+                        }
+                    }
+
+                }
+
+                // Remove {{ and }}
+                tokenSplits[0] = "";
+                tokenSplits[tokenSplits.length-1] = "";
+
+                // Reconstruct token
+                tokens[i] = TemplatePrinterUtils.join(tokenSplits);
+            }
+        }
+
+        // Reconstruct input
+        return TemplatePrinterUtils.join(tokens);
+    }
+
+    /**
+     * Validates the input string for well-formed {{ }} and < > pairs.
+     * If malformed, throws a RuntimeException which will be caught by
+     * doInBackground() and passed to the attached PopulateListener.onError()
+     * 
+     * @param input String to validate
+     */
+    private static void validateStringOrThrowException(String input) {
```

Comment:
Is it possible that we want to pass up some more information to the UI level like which pair was malformed or the approximate location?